source/material_model/averaging.cc

```diff
@@ -0,0 +1,449 @@
+/*
+  Copyright (C) 2011 - 2015 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include <deal.II/base/std_cxx11/array.h>
+#include <aspect/material_model/averaging.h>
+#include <deal.II/base/parameter_handler.h>
+#include <utility>
+#include <limits>
+
+using namespace dealii;
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    // Do the requested averaging operation for one array.
+    template <int dim>
+    void
+    Averaging<dim>::average (const std::string &averaging_operation,
+                             const std::vector<Point<dim> >    &position,
+                             std::vector<double>           &values_out) const
+    {
+      // if an output field has not been filled (because it was
+      // not requested), then simply do nothing -- no harm no foul
+      if (values_out.size() == 0)
+        return;
+
+      const unsigned int N = values_out.size();
+
+      // alfad is a constant which is dependent on the dimension and is used to define the shape of the bell shape.
+      const double alfad = (dim == 2 ? 5/(numbers::PI * bell_shape_limit * bell_shape_limit) : 106/(numbers::PI * bell_shape_limit * bell_shape_limit * bell_shape_limit));
+
+      // perform the requested averaging
+      if (averaging_operation == "none")
+        { }
+      else if (averaging_operation == "arithmetic average")
```

Comment:
you still do the string comparisons. please use the enum idea I mentioned previously -- string comparisons are so expensive that we don't want to do it on every cell if it can be avoided.